server/src/main/java/org/elasticsearch/injection/Planner.java

```diff
@@ -0,0 +1,144 @@
+     * @param requiredTypes the classes of which we need instances
+     * @param allParameterTypes the classes that appear as the type of any parameter of any constructor we might call
+     */
+    Planner(Map<Class<?>, InjectionSpec> specsByClass, Set<Class<?>> requiredTypes, Set<Class<?>> allParameterTypes) {
+        this.requiredTypes = requiredTypes;
+        this.plan = new ArrayList<>();
+        this.specsByClass = unmodifiableMap(specsByClass);
+        this.allParameterTypes = unmodifiableSet(allParameterTypes);
+        this.startedPlanning = new HashSet<>();
+        this.finishedPlanning = new HashSet<>();
+        this.alreadyProxied = new HashSet<>();
+
+        // Evolution note: this was a queue because we anticipated cases where the planner needs
+        // to put items at the end rather than recursing. If that never turns out to be necessary,
+        // this could be simplified away.
+        this.queue = new ArrayDeque<>(requiredTypes);
+    }
+
+    /**
+     * Intended to be called once.
+     * <p>
+     * Note that not all proxies are resolved once this plan has been executed.
+     * <p>
+     *
+     * <em>Evolution note</em>: in a world with multiple domains/subsystems,
+     * it will become necessary to defer proxy resolution until after other plans
+     * have been executed, because they could create additional objects that ought
+     * to be included in the proxies created by this plan.
+     *
+     * @return the {@link InjectionStep} objects listed in execution order.
+     */
+    List<InjectionStep> injectionPlan() {
+        Class<?> c;
+        while ((c = queue.poll()) != null) {
+            planForClass(c, 0);
+        }
+        return plan;
+    }
+
+    /**
+     * Recursive procedure that determines what effect <code>requestedClass</code>
+     * should have on the plan under construction.
+     *
+     * @param depth is used just for indenting the logs
+     */
+    private void planForClass(Class<?> requestedClass, int depth) {
+        InjectionSpec spec = specsByClass.get(requestedClass);
+        if (spec == null) {
+            throw new InjectionConfigurationException("Cannot instantiate " + requestedClass);
+        }
+        planForSpec(spec, depth);
+    }
+
+    private void planForSpec(InjectionSpec spec, int depth) {
+        String indent;
+        if (logger.isTraceEnabled()) {
+            indent = "\t".repeat(depth);
+        } else {
+            indent = null;
+        }
+
+        if (finishedPlanning.contains(spec)) {
+            logger.trace("{}Already planned {}", indent, spec);
+            return;
+        }
+
+        logger.trace("{}Planning for {}", indent, spec);
+        if (startedPlanning.add(spec) == false) {
+            // TODO: Better cycle detection and reporting. Use SCCs
+            throw new InjectionConfigurationException("Cyclic dependency involving " + spec);
+        }
+
+        if (spec instanceof MethodHandleSpec m) {
```

Comment:
I personally don't like this (not even when used with the new switch pattern matching); what about using a sort-of visitor pattern?

Reply:
IMHO the visitor pattern is strictly worse than type switches. Visitor is a way to implement type switches in languages that don't support them natively.

Reply:
The way to do it might be to have an `Enum` getter on the interface specifying its type that you can switch on. But if there's only ever going to be two types, it's probably not worth it.

Reply:
@ldematte - I don't mean that design patterns in general are for missing language features; only that Visitor specifically is unnecessary in languages with type switches, because the latter are strictly better.

Having said that, as someone who was really into the GoF Design Patterns back in the 90s (I had a photocopy of the summary pinned to my cubicle wall at my first job) I gradually became disillusioned with those particular patterns. Most of them are workarounds for missing language features of 90s-era C++, and many of the remainder are what I'd now consider antipatterns. (Their notion of Singleton is particularly poor, for example. The modern concept of singletons is much better.)

Reply:
@thecoop - How soon do we get to move to Java 21? In Java 21, we can switch on the types directly and wouldn't need the enum.

Reply:
I think it's planned for ES v9?